### PR TITLE
static analysis fixes

### PR DESCRIPTION
**Pull request by ebudai**

### CODE REVIEW

File: src/host/srvinit.cpp [hunk 1]
```diff
@@ -653,23 +653,23 @@ DWORD ConsoleIoThread()
...
-
-                // This will not return. Terminate immediately when disconnected.
-                ServiceLocator::RundownAndExit(STATUS_SUCCESS);
-            }
-            RIPMSG1(RIP_WARNING, "DeviceIoControl failed with Result 0x%x", hr);
-            ReplyMsg = nullptr;
-            continue;
-        }
+        
+			// TODO: 9115192 correct mixed NTSTATUS/HRESULT
```
Comment (binarycrusader) [inline src/host/srvinit.cpp]: I think you've got tabs here instead of spaces?

File: src/host/utf8ToWideCharParser.cpp [hunk 2]
```diff
@@ -325,8 +325,11 @@ unsigned int Utf8ToWideCharParser::_ParseFullRange(_In_reads_(cb) const byte* co
         {
             LOG_LAST_ERROR();
             _currentState = _State::Error;
-        }
-        _currentState = _State::Finished;
+		}
+		else 
+		{
+			_currentState = _State::Finished;
```
Comment (binarycrusader) [inline src/host/utf8ToWideCharParser.cpp]: tabs instead of spaces?

File: src/interactivity/win32/UiaTextRange.cpp [hunk 3]
```diff
@@ -466,7 +466,10 @@ IFACEMETHODIMP UiaTextRange::Compare(_In_opt_ ITextRangeProvider* pRange, _Out_
     }
     // tracing
     ApiMsgCompare apiMsg;
-    apiMsg.OtherId = other->GetId();
+	if (other)
```
Comment (binarycrusader) [inline src/interactivity/win32/UiaTextRange.cpp]: should be spaces instead of tabs?

File: src/renderer/base/thread.cpp [hunk 4]
```diff
@@ -41,13 +41,13 @@ RenderThread::~RenderThread()
     if (_hPaintEnabledEvent != INVALID_HANDLE_VALUE)
     {
         CloseHandle(_hPaintEnabledEvent);
-        _hEvent = INVALID_HANDLE_VALUE;
+		_hPaintEnabledEvent = INVALID_HANDLE_VALUE;
```
Comment (binarycrusader) [inline src/renderer/base/thread.cpp]: should be spaces, not tabs?

File: src/renderer/base/thread.cpp [hunk 5]
```diff
@@ -41,13 +41,13 @@ RenderThread::~RenderThread()
...
         CloseHandle(_hPaintEnabledEvent);
-        _hEvent = INVALID_HANDLE_VALUE;
+		_hPaintEnabledEvent = INVALID_HANDLE_VALUE;
     }
 
     if (_hPaintCompletedEvent != INVALID_HANDLE_VALUE)
     {
         CloseHandle(_hPaintCompletedEvent);
-        _hEvent = INVALID_HANDLE_VALUE;
+		_hPaintCompletedEvent = INVALID_HANDLE_VALUE;
```
Comment (binarycrusader) [inline src/renderer/base/thread.cpp]: should be spaces, not tabs?

File: src/server/ApiMessage.cpp [hunk 6]
```diff
@@ -153,7 +153,7 @@ HRESULT _CONSOLE_API_MSG::ReleaseMessageBuffers()
 
     if (State.InputBuffer != nullptr)
     {
-        delete[] State.InputBuffer;
+        delete[] (BYTE**) State.InputBuffer;
```
Comment (binarycrusader) [inline src/server/ApiMessage.cpp]: use C++ style casts

File: src/server/ApiMessage.cpp [hunk 7]
```diff
@@ -170,7 +170,7 @@ HRESULT _CONSOLE_API_MSG::ReleaseMessageBuffers()
             LOG_IF_FAILED(_pDeviceComm->WriteOutput(&IoOperation));
         }
 
-        delete[] State.OutputBuffer;
+        delete[] (BYTE*) State.OutputBuffer;
```
Comment (binarycrusader) [inline src/server/ApiMessage.cpp]: use C++ style casts

File: src/interactivity/win32/window.cpp [hunk 8]
```diff
@@ -239,133 +239,130 @@ NTSTATUS Window::_MakeWindow(_In_ Settings* const pSettings,
...
+        }
+
+        // Attempt to create window
+        HWND hWnd = CreateWindowExW(
+            CONSOLE_WINDOW_EX_FLAGS,
+            CONSOLE_WINDOW_CLASS,
+            gci.GetTitle().c_str(),
+            CONSOLE_WINDOW_FLAGS,
+            WI_IsFlagSet(gci.Flags,
+                            CONSOLE_AUTO_POSITION) ? CW_USEDEFAULT : rectProposed.left,
```
Comment (zadjii-msft) [inline src/interactivity/win32/window.cpp]: nit: this line needs to get de-indented too

File: src/renderer/vt/tracing.cpp [hunk 9]
```diff
@@ -43,7 +43,7 @@ std::string toPrintableString(const std::string_view& inString)
     for (size_t i = 0; i < inString.length(); i++)
     {
         unsigned char c = inString[i];
-        if (c < '\x20' && c < '\x7f')
```
Comment (zadjii-msft) [inline src/renderer/vt/tracing.cpp]: Actually, for this one, the `&& c < '\x7f'` clause should probably just be removed - the `c < '\x20' ` part was the important part.

File: src/server/ApiMessage.cpp (hunk 6 above)
Comment (zadjii-msft) [inline src/server/ApiMessage.cpp]: I'd be _really_ wary of changing this, the API code is notoriously fragile. Especially considering you're casting the two `delete[]`s to different types.

File: src/host/screenInfo.cpp [hunk 10]
```diff
@@ -1280,7 +1280,7 @@ void SCREEN_INFORMATION::_AdjustViewportSize(const RECT* const prcClientNew,
 
     const Viewport oldViewport = Viewport(_viewport);
 
-    _InternalSetViewportSize(pcoordSize, fResizeFromLeft, fResizeFromTop);
+    _InternalSetViewportSize(pcoordSize, fResizeFromTop, fResizeFromLeft);
```
Comment (zadjii-msft) [inline src/host/screenInfo.cpp]: great catch here - wonder how long that's been like this?

File: src/server/ApiMessage.cpp (hunk 6 above)
Comment (ebudai) [reply src/server/ApiMessage.cpp]: deleting void* is UB

File: src/renderer/vt/tracing.cpp (hunk 9 above)
Comment (ebudai) [reply src/renderer/vt/tracing.cpp]: If that is the case, then the else if (c == '\x7f') block can also be removed

Comment (ebudai) [reply src/renderer/vt/tracing.cpp]: Perhaps `&& c != '\x7f'` instead?

File: src/server/ApiMessage.cpp (hunk 6 above)
Comment (binarycrusader) [reply src/server/ApiMessage.cpp]: I believe @ebudai is right; in fact, in c++ 17 it's even more dangerous to abuse delete this way because deletion may need to meet specific alignment requirements originally set during new.  I also believe performing a delete on void won't cause destructors to be called.  The callers seems to always use byte, so I believe this to be correct (after it's changed to use C++ style casts).

Quoting from the draft standard:

http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/n4659.pdf

> 8.3.5 Delete
> ...
> In the first alternative (delete object), if the static type of the object to be deleted is different from its dynamic
> type, the static type shall be a base class of the dynamic type of the object to be deleted and the static type
> shall have a virtual destructor or the behavior is undefined. In the second alternative (delete array) if the
> dynamic type of the object to be deleted differs from its static type, the behavior is undefined.
> ...
> 82) This implies that an object cannot be deleted using a pointer of type void* because void is not an object type

File: src/renderer/vt/tracing.cpp (hunk 9 above)
Comment (ebudai) [reply src/renderer/vt/tracing.cpp]: Perhaps `&& c != '\x7f'` instead?

Comment (ebudai) [reply src/renderer/vt/tracing.cpp]: woops I was looking at `c > '\x20` the whole time.

File: src/host/ft_uia/KeyPressTests.cs [hunk 11]
```diff
@@ -208,7 +208,7 @@ public void VerifyCtrlCCmd()
                         for (int i = 0; i < text.Count(); ++i)
                         {
                             string line = text.ElementAt(i);
-                            line.Trim(' ');
+                            line = line.Trim(' ');
```
Comment (adiviness) [inline src/host/ft_uia/KeyPressTests.cs]: does the test still pass with this change?

File: src/host/utf8ToWideCharParser.cpp [hunk 12]
```diff
@@ -505,5 +508,5 @@ void Utf8ToWideCharParser::_Reset()
 {
     _currentState = _State::Ready;
     _bytesStored = 0;
-    _convertedWideChars.release();
+    delete _convertedWideChars.release();
```
Comment (adiviness) [inline src/host/utf8ToWideCharParser.cpp]: I would prefer `_convertedWideChars.reset(nullptr);` here instead of an explicit delete.

File: src/interactivity/win32/UiaTextRange.cpp [hunk 13]
```diff
@@ -466,7 +466,10 @@ IFACEMETHODIMP UiaTextRange::Compare(_In_opt_ ITextRangeProvider* pRange, _Out_
     }
     // tracing
     ApiMsgCompare apiMsg;
-    apiMsg.OtherId = other->GetId();
+    if (other)
+    {
+        apiMsg.OtherId = other->GetId();
+    }
```
Comment (adiviness) [inline src/interactivity/win32/UiaTextRange.cpp]: if `other` isn't there can this set `apiMsg.OtherId` to 0 (preferably with 0 being defined as some invalid id constant)? The static `UiaTextRange::id` would also need to be changed to start at 1 instead of 0, but that way we can still send the tracing message and detect when we got a bad api call.

Comment (ebudai) [reply src/interactivity/win32/UiaTextRange.cpp]: Where and how (static, macro, etc) to define the constant?

Comment (adiviness) [reply src/interactivity/win32/UiaTextRange.cpp]: `UiaTextRange::id` is defined near the top of UiaTextRange.hpp and it's initialized to 0 near the top of UiaTextRange.cpp. For the invalid id constant a `#define` in UiaTextRange.hpp would be fine.

File: src/host/ft_uia/KeyPressTests.cs (hunk 11 above)
Comment (ebudai) [reply src/host/ft_uia/KeyPressTests.cs]: yes

File: src/interactivity/win32/UiaTextRange.hpp [hunk 14]
```diff
@@ -29,6 +29,7 @@ Author(s):
 class UiaTextRangeTests;
 #endif
 
+#define INVALID_ID 0
```
Comment (dlong11) [inline src/interactivity/win32/UiaTextRange.hpp]: @ebudai INVALID_ID should be defined as a constexpr instead of a macro.  Take a look at Cpp Core guideline [ES.31](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es31-dont-use-macros-for-constants-or-functions)

File: src/host/srvinit.cpp [hunk 15]
```diff
@@ -653,22 +653,22 @@ DWORD ConsoleIoThread()
         if (ReplyMsg != nullptr)
         {
             LOG_IF_FAILED(ReplyMsg->ReleaseMessageBuffers());
-        }
```
Comment (miniksa) [inline src/host/srvinit.cpp]: I'm not so sure about this one. I think there are API calls that don't return a `ReplyMsg`. If any of those are called, then we would be stuck.

File: src/host/utf8ToWideCharParser.cpp [hunk 16]
```diff
@@ -505,5 +508,5 @@ void Utf8ToWideCharParser::_Reset()
 {
     _currentState = _State::Ready;
     _bytesStored = 0;
-    _convertedWideChars.release();
```
Comment (miniksa) [inline src/host/utf8ToWideCharParser.cpp]: I think this is also in #836 now

File: src/interactivity/win32/UiaTextRange.cpp [hunk 17]
```diff
@@ -20,7 +20,7 @@ using namespace Microsoft::Console::Interactivity::Win32::UiaTextRangeTracing;
 //#define UIATEXTRANGE_DEBUG_MSGS 1
 #undef UIATEXTRANGE_DEBUG_MSGS
 
-IdType UiaTextRange::id = 0;
```
Comment (miniksa) [inline src/interactivity/win32/UiaTextRange.cpp]: I want @adiviness to look at this specifically. UIA is fraught with 0 vs 1 things.

File: src/server/ApiMessage.cpp [hunk 18]
```diff
@@ -153,7 +153,7 @@ HRESULT _CONSOLE_API_MSG::ReleaseMessageBuffers()
 
     if (State.InputBuffer != nullptr)
     {
-        delete[] State.InputBuffer;
+        delete[] static_cast<BYTE**>(State.InputBuffer);
```
Comment (dreamer-dead) [inline src/server/ApiMessage.cpp]: Is `BYTE**` a correct type here?

File: src/host/srvinit.cpp (hunk 15 above)
Comment (ebudai) [reply src/host/srvinit.cpp]: In that case, with the old code, `ReplyMsg->Complete` would be a null dereference.

File: src/server/ApiMessage.cpp (hunk 18 above)
Comment (ebudai) [reply src/server/ApiMessage.cpp]: The only time InputBuffer is assigned is from `State.InputBuffer = pPayload.release();`, and `pPayload` is of type `wistd::unique_ptr<BYTE[]>`

File: src/interactivity/win32/UiaTextRange.hpp [hunk 19]
```diff
@@ -69,6 +67,8 @@ typedef unsigned int Column;
 // the first char of the 0th row in the text buffer row array.
 typedef unsigned int Endpoint;
 
+constexpr IdType invalid_id = 0;
```
Comment (dlong11) [inline src/interactivity/win32/UiaTextRange.hpp]: @adiviness What is the recommended naming scheme for constants?  I have seen all uppercase, pascal, lowercase, etc...   Which one is correct?

File: src/server/ApiMessage.cpp (hunk 18 above)
Comment (dreamer-dead) [reply src/server/ApiMessage.cpp]: `std::unique_ptr<T[]>` means that the pointer to the allocated memory will be deleted with `operator delete[]` by default. And the underlying type is still `T*`, in this case it is `BYTE*` and not `BYTE**`. See short sample program here https://ideone.com/P8kqXK

I'm not sure, but hope that `wistd:: unique_ptr` do the same thing.

Comment (ebudai) [reply src/server/ApiMessage.cpp]: You're right; I should have checked the return type for `pPayload.release()`.

File: src/host/srvinit.cpp (hunk 15 above)
Comment (ebudai) [reply src/host/srvinit.cpp]: Should be ok now that I see ReadIo() can take nullptr as 1st parameter.

File: src/interactivity/win32/UiaTextRange.cpp (hunk 17 above)
Comment (adiviness) [reply src/interactivity/win32/UiaTextRange.cpp]: this should be fine, I believe the id is used only for tracelogging.

File: src/interactivity/win32/UiaTextRange.hpp (hunk 19 above)
Comment (adiviness) [reply src/interactivity/win32/UiaTextRange.hpp]: It depends on the scope of the constant and what sort of constant it is. For macros we typically do all caps (ex. FOO_BAR). for constexprs we typically do pascal case (ex. FooBar).

Comment (binarycrusader) [reply src/interactivity/win32/UiaTextRange.hpp]: Relevant ISO C++ guideline:

[ES.9: Avoid ALL_CAPS names](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines.html#es9-avoid-all_caps-names)

TL;DR is that since ALL CAPS names are typically used for macros, it's dangerous to use them for anything but macros since it makes it easy for a macro to silently replace or redefine something that's not a macro.

Comment (dlong11) [reply src/interactivity/win32/UiaTextRange.hpp]: Thanks @binarycrusader - this is why I was asking.  I have seen a lot of constexpr that were upper case.  👎    I noticed an issue was already logged.  👍

File: src/host/srvinit.cpp (hunk 15 above)
Comment (miniksa) [reply src/host/srvinit.cpp]: You're right about the null dereference at `&ReplyMsg->Complete`. I looked back at the original code this was adapted from, and what should happen is:

1. ReadIo's signature should be taking the PCONSOLE_API_MSG ReplyMsg as parameter one and only attempting to dig out the completion part if it is not null at that level.
2. We need to keep the } on line 656 because there's technically a situation where the IO could fail for not pipe-disconnected and try to go around the loop again to reattempt the call with there being no reply message at all.
